driver-core/src/main/java/com/datastax/driver/core/HostConnectionPool.java

```diff
@@ -437,17 +461,23 @@ public int opened() {
         if (connections == null)
             return Lists.newArrayList(CloseFuture.immediateFuture());
 
-        List<CloseFuture> futures = new ArrayList<CloseFuture>(connections.size());
+        List<CloseFuture> futures = new ArrayList<CloseFuture>(connections.size() + trash.size());
+
         for (final PooledConnection connection : connections) {
             CloseFuture future = connection.closeAsync();
             future.addListener(new Runnable() {
                 public void run() {
-                    if (connection.markForTrash.compareAndSet(false, true))
+                    if (connection.state.compareAndSet(OPEN, GONE))
                         open.decrementAndGet();
                 }
             }, MoreExecutors.sameThreadExecutor());
             futures.add(future);
         }
+
+        // Some connections in the trash might still be open if they hadn't reached their idle timeout
+        for (PooledConnection connection : trash)
+            futures.add(connection.closeAsync());
```

Comment:
The point of GONE is to avoid a race between cleanup and "resurrection", which should not happen at shutdown (pending JAVA-599, which I plan to look at).